designs/serialization.md

```diff
@@ -0,0 +1,296 @@
+
+```python
+@runtime_checkable
+class ShapeSerializer(Protocol):
+
+    def begin_struct(
+        self, schema: "Schema"
+    ) -> AbstractContextManager["ShapeSerializer"]:
+        ...
+
+    def write_struct(self, schema: "Schema", struct: "SerializeableStruct") -> None:
+        with self.begin_struct(schema=schema) as struct_serializer:
+            struct.serialize_members(struct_serializer)
+
+    def begin_list(self, schema: "Schema") -> AbstractContextManager["ShapeSerializer"]:
+        ...
+
+    def begin_map(self, schema: "Schema") -> AbstractContextManager["MapSerializer"]:
+        ...
+
+    def write_null(self, schema: "Schema") -> None:
+        ...
+
+    def write_boolean(self, schema: "Schema", value: bool) -> None:
+        ...
+
+    def write_byte(self, schema: "Schema", value: int) -> None:
+        self.write_integer(schema, value)
+
+    def write_short(self, schema: "Schema", value: int) -> None:
+        self.write_integer(schema, value)
+
+    def write_integer(self, schema: "Schema", value: int) -> None:
+        ...
+
+    def write_long(self, schema: "Schema", value: int) -> None:
+        self.write_integer(schema, value)
+
+    def write_float(self, schema: "Schema", value: float) -> None:
+        ...
+
+    def write_double(self, schema: "Schema", value: float) -> None:
+        self.write_float(schema, value)
+
+    def write_big_integer(self, schema: "Schema", value: int) -> None:
+        self.write_integer(schema, value)
+
+    def write_big_decimal(self, schema: "Schema", value: Decimal) -> None:
+        ...
+
+    def write_string(self, schema: "Schema", value: str) -> None:
+        ...
+
+    def write_blob(self, schema: "Schema", value: bytes) -> None:
+        ...
+
+    def write_timestamp(self, schema: "Schema", value: datetime.datetime) -> None:
+        ...
+
+    def write_document(self, schema: "Schema", value: "Document") -> None:
+        ...
+
+
+@runtime_checkable
+class MapSerializer(Protocol):
+    def entry(self, key: str, value_writer: Callable[[ShapeSerializer], None]):
+        ...
+
+
+@runtime_checkable
+class SerializeableShape(Protocol):
+    def serialize(self, serializer: ShapeSerializer) -> None:
+        ...
+
+
+@runtime_checkable
+class SerializeableStruct(SerializeableShape, Protocol):
+    def serialize_members(self, serializer: ShapeSerializer) -> None:
+        ...
+```
+
+Below is an example Smithy `structure` shape, followed by the
+`SerializebleShape` it would generate.
+
+```smithy
+namespace com.example
+
+structure ExampleStructure {
+    member: Integer = 0
+}
+```
+
+```python
+@dataclass(kw_only=True)
+class ExampleStructure:
+    member: int = 0
+
+    def serialize(self, serializer: ShapeSerializer):
+        serializer.write_struct(EXAMPLE_STRUCTURE_SCHEMA, self)
+
+    def serialize_members(self, serializer: ShapeSerializer):
+        serializer.write_integer(
+            EXAMPLE_STRUCTURE_SCHEMA.members["member"], self.member
```

Comment:
To avoid any hashmap/hashing overhead, I would either use the member index or also generate constants that refer to each member

Reply:
The member index can't be used as it refers to the order present in the model at generation time, which is not the same as the order in the members dict. Generating constants for each member would add massive bloat to the artifact size, which I think offsets whatever gains you might get in performance.

Reply:
You have to add, at minimum, two extra lines of code for *every member* - one to define and and one to import it. We might end up switching to using a wildcard import though, which would eliminate the lines bit.